apps/vaporgui/PVariableSelector.cpp

```diff
@@ -0,0 +1,26 @@
+#include "PVariableSelector.h"
+#include "VCheckBox.h"
+#include "VLineItem.h"
+#include <vapor/ParamsBase.h>
+#include <vapor/RenderParams.h>
+#include <assert.h>
+
+using VAPoR::RenderParams;
+using VAPoR::Box;
+
+PVariableSelector::PVariableSelector(const std::string &tag, const std::string &label)
+: PStringDropdown(tag, {}, label)
+{}
+
+void PVariableSelector::updateGUI() const
+{
+    RenderParams *rp = dynamic_cast<RenderParams*>(getParams());
+    assert(rp && "Params must be RenderParams");
+    
+    int nDims = rp->GetBox()->GetOrientation() == Box::XY ? 2 : 3;
```

Comment:
This is not going to fly for variable dimension selection. There may be instances where a renderer mixes 2D and 3D variables for one thing. Perhaps the constructor should require a pointer to the method used to get the currently selected variable from the params data base, and then use the DataMgr to determine that variables dimensionality.  Note that most renderers strictly support either 2D or 3D variables, not both.

Reply:
Sure but does this need to support every possible use case? Currently we manually code/configure every variable selector. This can take care of 90% of cases and the last 10% we can continue hard-coding as we do now. It could make sense to have the variable dimension to be 2D, 3D, or auto.

Reply:
Which use cases would it not support? It's not clear how you would "manually configure" this. The salient thing is we should be detecting the dimensionality by assuming that its dimensionality matches that of the *default* Box class.

Reply:
I don't know but I think it can support most of them without requiring a function pointer to initialize the dimensionality. Also that would not work in the case of the Flow or Wireframe renderers which change their dimensionality. I say the constructor creates a selector which always shows 2D vars, 3D vars, or automatically selects the var based on the determined dimension of the renderer. So for example:

Contour renderer:
```
VarSel("main", Auto);
VarSel("height", 2D);
```

Flow renderer:
```
VarSel("X", Auto);
VarSel("Y", Auto);
VarSel("Z", Auto);
VarSel("Color", Auto);
VarSel("Height", 2D);
```

If you have any more specific edge case rather than adding support for function pointers to allow VariableSelector to support everything just don't use it and do what we currently do or extend it.

Reply:
I was focused on the issue with using the Box to determine the dimensionality and it just occurred to me that you don't have a "HLI" interface to the PVariableWidget. Is there a reason for this? Once implemented, isn't it simply a matter of calling the Params getter method for the variable and then querying the variable's dimensionality via the DataMgr?

Reply:
What would be the correct way to get the dimensionality? I tested various methods and this seemed to be the most reliable. Querying the dimensionality of the currently selected variable does not seem like a better solution. For example many variables slots have no variable as a valid option, e.g. the height variable or the Z variable for the flow or barbs. The reason there is no HLI is because I wrote this class as a specialized element that I've been using in my code, not a generic building block. If you need to call a specific function, for example to set the main variable for a renderer, a better solution would be to create a subclass, PMainVariableSelector, or something of that sort that can be used whenever you need to set the main variable rather than passing in the setter and getter functions for RenderParams::SetVariable for every renderer's GUI. If someone really needs an HLI version for some reason they can just call `CreateHLI(PVariableSelector,  std::string);`

Reply:
First, determining the variable dimensionality should only be an issue for the "Auto" configuration, correct? If the selector is configured 2D or 3D there is no need to figure out the dimensionality of the currently selected variable because it can't change.

In the case that the selector is configured "Auto" the dimensionality should be determined by querying the dimensionality of the currently selected variable using DataMgr:GetNumDimensions(). If there is no currently selected variable and the configuration is "Auto" then the choice is arbitrary. 

Relying on the dimensionality of the default Box returned by RenderParams::GetBox() might be working for you currently, but there is no guarantee that it will continue to work in the future.

Reply:
Ok. Done.

Reply:
What happens in

`return getDataMgr()->GetNumDimensions(getParamsString());`

when 

`getParamsString()`

returns an empty value?

Also, you need to update the class documentation for PVariableSelector; it still references GetBox()

Reply:
It returns 0 as the number of dimensions so we get an empty drop down. I think this is a better option than it arbitrarily choosing 2D or 3D as then we may see other bugs caused by cases such as renderers getting 2D variables when they expect 3D or vice versa. I fixed the documentation.